### PR TITLE
oadp-1.3: OADP-4265 Mark InProgress backup/restore as failed upon requeuing

### PR DESCRIPTION
**Pull request by kaovilai**
Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes #(issue)

# Please indicate you've done the following:

- [ ] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

**Comment by kaovilai** [comment]
from prior feedback need to do finalizer (operations in 1.3) controller also

**Comment by kaovilai** [comment]
> from prior feedback need to do finalizer (operations in 1.3) controller also

For restore operations controller we're covered by this returning error: https://github.com/openshift/velero/blob/c06b13825cd7d2df1ea3b70d0df725006127d92e/pkg/controller/restore_operations_controller.go#L194

backup operations controller: https://github.com/openshift/velero/blob/c06b13825cd7d2df1ea3b70d0df725006127d92e/pkg/controller/backup_operations_controller.go#L207

**Comment by sseago** [comment]
Yes, I think we're fine leaving the backup/restore_operations_controller returns alone, since those already requeue. I don't think we have restore finalizer in 1.3, so we just need backup, restore, and backup finalizer.

**Comment by kaovilai** [comment]
If patch fails on finalizer controller we wanna retry patch as fail? It's kinda difficult there since it's currently using defer func.

I would have to break the patch call out of defer to return reconciler err on patch fail. Is that ok with you?

File: pkg/controller/backup_controller.go [hunk 1]
```diff
@@ -233,6 +233,20 @@ func (b *backupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
 	case "", velerov1api.BackupPhaseNew:
 		// only process new backups
+	case velerov1api.BackupPhaseInProgress:
+		// if backup is in progress, we should not process it again
+		// we want to mark it as failed to avoid it being stuck in progress
+		// if so, mark it as failed, last loop did not successfully complete the backup
+		log.Debug("Backup has in progress status from prior reconcile, marking it as failed")
+		failedCopy := original.DeepCopy()
+		failedCopy.Status.Phase = velerov1api.BackupPhaseFailed
+		failedCopy.Status.FailureReason = "Backup from previous reconcile still in progress"
```
Comment (sseago) [inline pkg/controller/backup_controller.go]: We may want to suggest an APIServer failure here.
"Backup from previous reconcile still in progress. The API Server may have been down."

File: pkg/controller/backup_controller.go [hunk 2]
```diff
@@ -249,7 +263,6 @@ func (b *backupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 		request.Status.Phase = velerov1api.BackupPhaseInProgress
 		request.Status.StartTimestamp = &metav1.Time{Time: b.clock.Now()}
 	}
-
```
Comment (sseago) [inline pkg/controller/backup_controller.go]: To minimize change from upstream since this is in our fork, lets not include whitespace changes like this.

File: pkg/controller/restore_controller.go [hunk 3]
```diff
@@ -197,29 +197,40 @@ func (r *restoreReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
 	case "", api.RestorePhaseNew:
 		// only process new restores
+	case api.RestorePhaseInProgress:
+		// if restore is in progress, we should not process it again
+		// we want to mark it as failed to avoid it being stuck in progress
+		// if so, mark it as failed, last loop did not successfully complete the restore
+		log.Debug("Restore has in progress status from prior reconcile, marking it as failed")
+		failedCopy := original.DeepCopy()
+		failedCopy.Status.Phase = api.RestorePhaseFailed
+		failedCopy.Status.FailureReason = "Restore from previous reconcile still in progress"
```
Comment (sseago) [inline pkg/controller/restore_controller.go]: We may want to suggest an APIServer failure here.
"Restore from previous reconcile still in progress. The API Server may have been down."

File: pkg/controller/restore_controller.go [hunk 4]
```diff
@@ -162,8 +162,8 @@ func (r *restoreReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 	// the controller.
 	log := r.logger.WithField("Restore", req.NamespacedName.String())
 
-	restore := &api.Restore{}
-	err := r.kbClient.Get(ctx, client.ObjectKey{Namespace: req.Namespace, Name: req.Name}, restore)
+	original := &api.Restore{}
```
Comment (sseago) [inline pkg/controller/restore_controller.go]: Is the variable name change necessary here? This makes the diff larger and increases the possibility of rebase conflicts, since we're carrying this commit in our fork.

**Review by sseago — CHANGES_REQUESTED** [review]
A couple of minor comments on error messages and the change of variable name. The bigger issue is we need the change for the backup finalizer controller as well (no restore finalizer controller in Velero 1.12, so that's not a concern here).

**Comment by sseago** [comment]
> If patch fails on finalizer controller we wanna retry patch as fail? It's kinda difficult there since it's currently using defer func.
> 
> I would have to break the patch call out of defer to return reconciler err on patch fail. Is that ok with you?

Hmm. I think we need to do this in some way. If we can't do it with defer, then we'll need to eliminate the defer call and include this with all return statements.

File: pkg/controller/backup_finalizer_controller.go [hunk 5]
```diff
@@ -106,15 +106,21 @@ func (r *backupFinalizerReconciler) Reconcile(ctx context.Context, req ctrl.Requ
...
-		case velerov1api.BackupPhaseCompleted, velerov1api.BackupPhasePartiallyFailed, velerov1api.BackupPhaseFailed, velerov1api.BackupPhaseFailedValidation:
-			r.backupTracker.Delete(backup.Namespace, backup.Name)
-		}
 		// Always attempt to Patch the backup object and status after each reconciliation.
-		if err := r.client.Patch(ctx, backup, kbclient.MergeFrom(original)); err != nil {
-			log.WithError(err).Error("Error updating backup")
-			return
+		if !patched {
+			if err := r.client.Patch(ctx, backup, kbclient.MergeFrom(original)); err != nil {
+				log.WithError(err).Error("Error updating backup")
```
Comment (mateusoliveira43) [inline pkg/controller/backup_finalizer_controller.go:115]: should we update `patchErr` here and not return?

File: pkg/controller/backup_finalizer_controller.go
Comment (mateusoliveira43) [inline pkg/controller/backup_finalizer_controller.go:1]: we are not doing the same thing for restores, because it does not exist a `restore_finalizer_controller`, right?

File: pkg/controller/backup_finalizer_controller.go (hunk 5 above)
Comment (mateusoliveira43) [reply pkg/controller/backup_finalizer_controller.go:115]: does not make sense, it would be true and would not trigger following if

resolving

File: pkg/controller/restore_controller.go (hunk 4 above)
Comment (kaovilai) [reply pkg/controller/restore_controller.go]: I did this so restore_controller and backup_controller has the same var name pattern that's all.

Comment (kaovilai) [reply pkg/controller/restore_controller.go]: Not necessary I agree, just make the logic more pastable across both.

File: pkg/controller/backup_finalizer_controller.go
Comment (sseago) [reply pkg/controller/backup_finalizer_controller.go:1]: That's right. Restore finalizer controller is in Velero 1.14, but not in 1.12

File: pkg/controller/restore_controller.go (hunk 4 above)
Comment (sseago) [reply pkg/controller/restore_controller.go]: @kaovilai I figured that's why you did that. I reverted that part here in a later commit since we're carrying the commit in our fork right now, and it removed a lot of lines from the diff, making rebase conflicts less likely. Does that seem reasonable to you?

Comment (kaovilai) [reply pkg/controller/restore_controller.go]: yes. That's reasonable

**Comment by kaovilai** [comment]
lgtm

**Comment by weshayutin** [comment]
ok.. testing update: 

w/o the patch, the backup stayed in progress.  While updating the dpa the velero server was restarted.

backup: westest-vsphere-apidown-1

```
status:
  completionTimestamp: "2024-06-17T17:25:58Z"
  expiration: "2024-07-17T17:12:21Z"
  failureReason: found a backup with status "InProgress" during the server starting,
    mark it as "Failed"
```
1.3.0 have to change the csv to get the test image on:

```
oc get csv oadp-operator.v1.3.0 -o yaml | grep sseago
                  value: quay.io/sseago/velero:1.3-requeue
```
Once patched, I initiated a second backup and then took down the api server for roughly 1 minute:

backup: westest-vsphere-apidown-2

```
status:
  expiration: "2024-07-17T17:38:01Z"
  failureReason: Backup from previous reconcile still in progress. The API Server
    may have been down.
  formatVersion: 1.1.0
  phase: Failed
  startTimestamp: "2024-06-17T17:38:01Z"
  version: 1
```

**Review by weshayutin — APPROVED** [review]
/LGTM

**Review by sseago — APPROVED** [review]
Reviewing to remove my prior "changes requested", but don't count it as an ack, since my own changes are in here as well.

**Comment by mateusoliveira43** [comment]
Lets trying modifying here https://github.com/openshift/velero/blob/oadp-1.3/Makefile.prow#L61 from latest to v0.17.5 to see if CI passes?

**Comment by kaovilai** [comment]
Just noting that this may never upstream based on comments at https://github.com/vmware-tanzu/velero/pull/7863#discussion_r1636304591

File: Makefile.prow [hunk 6]
```diff
@@ -58,7 +58,7 @@ envtest: $(GOBIN)/setup-envtest
 
 $(GOBIN)/setup-envtest:
 	@echo Installing envtest tools
-	GOFLAGS= go install sigs.k8s.io/controller-runtime/tools/setup-envtest@latest
+	GOFLAGS= go install sigs.k8s.io/controller-runtime/tools/setup-envtest@v0.0.0-20230216140739-c98506dc3b8e
```
Comment (sseago) [inline Makefile.prow]: This really needs to be in a separate PR. Since this PR is for velero changes that aren't (yet?) upstream, if they ever did make it upstream, we'd be dropping this commit, and then we'd lose the prow fix.

Comment (kaovilai) [reply Makefile.prow]: so merge that before this?

Comment (sseago) [reply Makefile.prow]: I think so, yes. Also, won't we need that same fix elsewhere (although maybe it's already on 1.4 and konveyor-dev). Just trying to keep carried patches from upstream code limited to just those changes.

Comment (kaovilai) [reply Makefile.prow:61]: [will drop commit once test passes and open new pr](https://github.com/openshift/velero/pull/320)

Comment (kaovilai) [reply Makefile.prow]: https://github.com/openshift/velero/pull/320

Comment (kaovilai) [reply Makefile.prow]: reverted.

**Comment by kaovilai** [comment]
retest after https://github.com/openshift/velero/pull/320

**Comment by kaovilai** [comment]
follow on bugfix: https://github.com/openshift/velero/pull/324
